File: Assets/Synth/Oscillator.js
#pragma strict

class Oscillator {
	private var cx = 0.0;
	private var step = 0.0;

	static private var kPi2 = 6.28318530718;
	
	function SetNote(note : int) {
		var freq = 440.0 * Mathf.Pow(2.0, 1.0 * (note - 69) / 12.0);
		step = freq / SynthConfig.kSampleRate;
	}
	
	function Run() {
		cx += step;
		cx -= Mathf.Floor(cx);
		return Mathf.Sin(kPi2 * cx);
	}
}

File: Assets/Synth/Oscillator.cs
namespace Arpie.Synth {

struct Oscillator
{
    const float Pi2 = 6.28318530718f;

    static float NoteFreq(int note)
      => 440 * UnityEngine.Mathf.Pow(2, (note - 69) / 12.0f);

    float _step;
    float _phase;

    static public Oscillator Note(int note)
      => new Oscillator { _step = NoteFreq(note) / Config.SampleRate };

    public float Run()
    {
        _phase += _step;
        _phase -= UnityEngine.Mathf.Floor(_phase);
        return UnityEngine.Mathf.Sin(Pi2 * _phase);
    }
}

} // namespace Arpie.Synth
